compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -732,6 +783,23 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                         let principal_a = data_a.principal().unwrap();
                         let target_trait_did = principal_def_id_b.unwrap();
                         let source_trait_ref = principal_a.with_self_ty(self.tcx(), source);
+                        if self.need_migrate_deref_output_trait_object(
+                            source,
+                            &obligation.cause,
+                            obligation.param_env,
+                        ) == Some(target_trait_did)
+                        {
+                            self.tcx().struct_span_lint_hir(
+                                DEREF_INTO_DYN_SUPERTRAIT,
+                                obligation.cause.body_id,
+                                obligation.cause.span,
+                                |lint| {
+                                    lint.build("use of the a `Deref` implementation with a `dyn SuperTrait` type as `Output`").emit();
```

Comment:
Maybe the names of the traits in question should be included in the message?

```rust
lint.build(format!(
    “`{}` implements `Deref` with supertrait `{}` as output”,
    source,
    target
))
```

Which should give errors like this:

    warning: `dyn B` implements `Deref` with supertrait `dyn A` as output
       --> $DIR/unsupported.rs:39:1
         |
      LL |     take_a(b);
         |     ^^^^^^^^^
         |
         = note: `#[warn(deref_into_dyn_supertrait)]` on by default
         = warning: this was previously accepted by the compiler but is being phased out;
                    it will become a hard error in a future release!
         = note: for more information, see issue ...

Reply:
Done.